File: 424. Longest Repeating Character Replacement/424. Longest Repeating Character Replacement.java
class Solution {

  public int characterReplacement(String s, int k) {
    int[] alphabet = new int[26];
    int maxLength = 0;
    int maxCount = 0;
    int left = 0;

    for (int i = 0; i < s.length(); i++) {
      int alphabetIndex = s.charAt(i) - 'A';
      alphabet[alphabetIndex]++;
      maxCount = Math.max(maxCount, alphabet[alphabetIndex]);

      while (i - left + 1 - maxCount > k) {
        alphabet[s.charAt(left) - 'A']--;
        left++;
      }

      maxLength = Math.max(maxLength, i - left + 1);
    }

    return maxLength;
  }
}

File: 424. Longest Repeating Character Replacement/424. Longest Repeating Character Replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function (s, k) {
  const alphabet = new Array(26).fill(0);
  let maxLength = 0;
  let maxCount = 0;
  let left = 0;

  for (let i = 0; i < s.length; i++) {
    const alphabetIndex = s.charCodeAt(i) - "A".charCodeAt(0);
    alphabet[alphabetIndex]++;
    maxCount = Math.max(maxCount, alphabet[alphabetIndex]);

    while (i - left + 1 - Math.max(...alphabet) > k) {
      alphabet[s.charCodeAt(left) - "A".charCodeAt(0)]--;
      left++;
    }

    maxLength = Math.max(maxLength, i - left + 1);
  }

  return maxLength;
};
